### ISSUE TITLE
[Regression] 0.19.0 fails to start while starting my X server

### ISSUE THREAD

**Issue by albert-astals-cid-kdab** [sddm/sddm#1316]
This is what systemctl status stays

```
● sddm.service - Simple Desktop Display Manager
     Loaded: loaded (/usr/lib/systemd/system/sddm.service; enabled; vendor preset: disabled)
     Active: active (running) since Thu 2020-11-05 18:09:49 CET; 40s ago
       Docs: man:sddm(1)
             man:sddm.conf(5)
   Main PID: 471 (sddm)
      Tasks: 2 (limit: 38081)
     Memory: 23.2M
     CGroup: /system.slice/sddm.service
             └─471 /usr/bin/sddm

de nov. 05 18:09:49 xps15 sddm[471]: Initializing...
de nov. 05 18:09:49 xps15 sddm[471]: Starting...
de nov. 05 18:09:49 xps15 sddm[471]: Logind interface found
de nov. 05 18:09:50 xps15 sddm[471]: Adding new display on vt 1 ...
de nov. 05 18:09:50 xps15 sddm[471]: Loading theme configuration from ""
de nov. 05 18:09:50 xps15 sddm[471]: Display server starting...
de nov. 05 18:09:50 xps15 sddm[471]: Adding cookie to "/var/run/sddm/{d02df92c-de3d-4dfc-84b9-1f563812434a}"
de nov. 05 18:09:50 xps15 sddm[471]: Running: /usr/bin/X -nolisten tcp -auth /var/run/sddm/{d02df92c-de3d-4dfc-84b9-1f563812434a} -background none -noreset -displayfd 17 -seat seat0 vt1
de nov. 05 18:09:50 xps15 sddm[471]: Failed to read display number from pipe
de nov. 05 18:09:50 xps15 sddm[471]: Could not start Display server on vt 1
```

I've traced the regression to commit 74394fd1cff26b0516af87539906a31bf2f49825

This is most probably a race since i can 100% reproduce the issue on my fast i9 but on the older i7 and i5 it works fine.

Another thing that points into a race is that if i going into vt 2, log in and issue a `systemctl restart sddm`, it brings sddm up just fine

**Comment by albert-astals-cid-kdab** [comment]
@davidedmundson @antonio-rojas since i've talked with you both in private about this, tagging you both here

**Comment by albert-astals-cid-kdab** [comment]
Further indication that there seems to be a race, https://github.com/tsdgeos/sddm/commit/00bef202a79df49a4a8ba579da3ad0fe6f58a190 "fixes" it :D

**Comment by nowshed-imran** [comment]
Facing the same problem here. Switched to `lightdm`.

**Comment by milianw** [comment]
I also ran into this on a pretty fast workstation. for me it's not always happening, which also seems to indicate it's a (random) race. 

thankfully the workaround mentioned by @albert-astals-cid-kdab works reliably.

**Comment by belavizer** [comment]
Confirm the same with mesa (20.2.1-1),  xf86-video-amdgpu (19.1.0-2), linux (5.9.4.arch1-1) using sddm-0.19.0-2-x86_64 from Arch Linux. I downgraded to sddm-0.18.1-3-x86_64 and looks it is ok now.

**Comment by davidedmundson** [comment]
Can someone add

[X11] 
ServerArguments="-logverbose 6"

to their config

Hopefully X will tell us something

**Comment by albert-astals-cid-kdab** [comment]
[xorg.no.txt](https://github.com/sddm/sddm/files/5505564/xorg.no.txt)
[xorg.yes.txt](https://github.com/sddm/sddm/files/5505565/xorg.yes.txt)

So when it fails is because it can't find my intel card yet?

**Comment by Vogtinator** [comment]
SDDM relies on the `CanGraphical` property of the systemd seat, which is set by udev rules in /usr/lib/udev/rules.d/71-seat.rules (`TAG+="master-of-seat"`) for several types of output devices. This is only set once any DRM device appears, but if `nomodeset` set also applies to plain framebuffers.

What probably happens here is that the `master-of-seat` device is somehow not enough for X to work. Without 74394fd, this made `sddm` crash and systemd restarted its service.

Unfortunately I don't know what could be done about this, other than maybe attempting to start the display a second time if the first one failed.

**Comment by tsdgeos** [comment]
That's a possible solution, one could look at what gdm and lightdm do since those seem to work fine here.

On the other hand, why does sddm say "i'm all good" when it failed to start X, that's not being "all good", that's having failed. If you instead say "i failed", systemd will take care of restarting sddm again and then it will "work" just fine, no?

**Comment by nowshed-imran** [comment]
Maybe I am over stepping the boundaries here. 
However, core developers can consider integrating SDDM under KDE project banner while explicitly independent of KDE libraries.
That way, more developers can invest on this crucial project. 
Until recently this project was basically stale. 
This is a suggestion from a KDE desktop lover who considers SDDM to be a very important piece of software.

**Comment by davidedmundson** [comment]
>So when it fails is because it can't find my intel card yet?

That's very much what it looks like

I went through lightdm code again, I can't see it doing anything different to us in terms of waiting before the initial start.

**Comment by davidedmundson** [comment]
This seems somewhat relevant: https://github.com/systemd/systemd/issues/10435 ?

It's not quite the same, but it's the idea that not all devices are set up for a seat when CanGraphical changes.

@tsdgeos can I have a full journald log of the bootup. From a working session should suffice, so you don't have to logout.

**Comment by tsdgeos** [comment]
Can you give me the exact command you want me to type to get the full journald log of the bootup? I've no idea what those words mean ^_^

**Comment by Vogtinator** [comment]
> On the other hand, why does sddm say "i'm all good" when it failed to start X, that's not being "all good", that's having failed. If you instead say "i failed", systemd will take care of restarting sddm again and then it will "work" just fine, no?

SDDM can manage multiple displays and seats, including multiple X servers. You probably don't want that all sessions die if X fails to start after you click on `Switch User` in Plasma...

> However, core developers can consider integrating SDDM under KDE project banner while explicitly independent of KDE libraries.
That way, more developers can invest on this crucial project.

I don't think that would make any difference. Anyone can make a PR, whether that's on github.com or invent.kde.org.

> This seems somewhat relevant: systemd/systemd#10435 ?

Yes, but only X itself can actually say when it's ready to start. There's no way around that, without knowing which devices will appear in the future (which needs a time machine).

> Can you give me the exact command you want me to type to get the full journald log of the bootup? I've no idea what those words mean ^_^

`journalctl -b --no-pager > file`

**Comment by albert-astals-cid-kdab** [comment]
Here it comes 
[journalctl.log.txt](https://github.com/sddm/sddm/files/5508980/journalctl.log.txt)

**Comment by Vogtinator** [comment]
Yep, that confirms it:

```
de nov. 09 09:21:26 xps15 kernel: [drm] Initialized nvidia-drm 0.0.0 20160202 for 0000:01:00.0 on minor 1
[...]
de nov. 09 09:21:26 xps15 sddm[500]: Running: /usr/bin/X -logverbose 6 -auth /var/run/sddm/{c4348a38-4a35-457f-9622-212470d58dd1} -background none -noreset -displayfd 17 -seat seat0 vt1
[...]
de nov. 09 09:21:27 xps15 kernel: [drm] Initialized i915 1.6.0 20200715 for 0000:00:02.0 on minor 0
```

Please try this, it's a bit cleaner as a workaround: https://github.com/sddm/sddm/pull/1324

While it's still ugly, I can't think of a better approach and the benefit of this is that it won't make any difference to systems where it's already working.

**Comment by abucodonosor** [comment]
Hello,

I have the same issue on all my Laptops. I don't have Nvidia or anything like this, All laptops have AMD APU's with and without
AMD RX xxxx dGPUs.

For me reverting https://github.com/sddm/sddm/commit/74394fd1cff26b0516af87539906a31bf2f49825 fixed it.

Best Regards.

**Comment by nowshed-imran** [comment]
@Vogtinator If this fixes the bug kindly consider making 0.19.1 release. 
This bug is critical for my system.

**Comment by kaktuspalme** [comment]
I'm not sure if my findings are related to this but I couldn't get to the plasma desktop too after updating to 0.19.

I have autologin activated, so I reverted back to 0.18.1 and deactivated autologin there. Updated to 0.19 again, activated autologin and it worked. When I compared the old /etc/sddm.conf.d/kde_settings.conf from a backup with the new one, the line "Session=plasma" was missing in the autologin section in the file which didn't work with 0.19.

As soon as this line is missing I can't get to the desktop.

What may be also interesting, when it's not working I have this line in the logs:
`sddm-helper[1207]: Starting: "/usr/share/sddm/scripts/wayland-session /usr/bin/startplasma-x11"`
I don't use wayland and when Session=plasma is in kde_settings.conf this line is missing.

**Comment by Vogtinator** [comment]
> I'm not sure if my findings are related to this but I couldn't get to the plasma desktop too after updating to 0.19.

Probably a different issue. If it was this bug, then `systemctl restart display-manager` would fix it.

> I have autologin activated, so I reverted back to 0.18.1 and deactivated autologin there. Updated to 0.19 again, activated autologin and it worked. When I compared the old /etc/sddm.conf.d/kde_settings.conf from a backup with the new one, the line "Session=plasma" was missing in the autologin section in the file which didn't work with 0.19.
> 
> As soon as this line is missing I can't get to the desktop.
>
> What may be also interesting, when it's not working I have this line in the logs:
> `sddm-helper[1207]: Starting: "/usr/share/sddm/scripts/wayland-session /usr/bin/startplasma-x11"`
> I don't use wayland and when Session=plasma is in kde_settings.conf the above line is missing in the logs.

That means that the default autologin session is `plasmawayland.desktop`. That might've been caused by https://github.com/sddm/sddm/pull/1305. There's nothing we can do on the sddm side that the default session doesn't work, it's user configurable.

**Comment by kaktuspalme** [comment]
@Vogtinator Thank you for explanation, restarting display manager doesn't fix it. So it's not related to this. I'll report it back. Thanks

**Comment by DataDrake** [comment]
@Vogtinator I've been doing some digging into this with @kaktuspalme and it looks like the following is happening:

**Pre-conditions**
1. `Session` is empty or missing for `[Autologin]` in the SDDM `mainConfig`
2. `lib/sddm/state.conf` has been written out under 0.18.1 with an absolute path (ie. `/usr/share/xsessions/plasma.desktop`)

**Autologin Process**

https://github.com/sddm/sddm/blob/46f08fb8c61bda7fa0e12fa022d058f54d351d3a/src/daemon/Display.cpp#L108-L121

1. `autologinSession` is found to be empty
2. `autologinSession` is overwritten with the last session from `lib/sddm/state.conf`
3. `findSessionEntry` is called to check for a valid Wayland session:

https://github.com/sddm/sddm/blob/46f08fb8c61bda7fa0e12fa022d058f54d351d3a/src/daemon/Display.cpp#L247-L256

4. `dir.exists()` is called with `/usr/share/xsessions/plasma.desktop` which has an absolute path.
5. Per [the docs](https://doc.qt.io/qt-5/qdir.html#exists) the absolute path will return true if it exists, even if it isn't in the path specified by the QDir.
6. The session type is switched to Wayland.
7. The session is run as Wayland with an Xsession `.desktop`.
8. Autologin fails.

Prior to [this commit](https://github.com/sddm/sddm/commit/c78da43cf44a1090d8aa1d5ed4a779a3ec42c504), X11 was being evaluated first to the opposite effect. The absolute path would override the relative check and the session would continue on without failing, being correctly set to X11 instead of Wayland.

It seems to me that this could be fixed if one or both of these changes were made:
1. Absolute paths in `Session` from `var/lib/state.conf` should be treated differently than relative paths.
2. `var/lib/state.conf` should not be written out with an absolute path, possibly by also writing out a `SessionDir` field to preserve the full path.

I'm happy to provide any additional info you might need.

**Comment by Vogtinator** [comment]
Can you please move that into a new issue?

**Comment by DataDrake** [comment]
@Vogtinator Sure thing

**Comment by nowshed-imran** [comment]
The problem is fixed in Archlinux. 
Thanks!

**Comment by YamiYukiSenpai** [comment]
I just ran into this after I swapped my GPU.

I had to downgrade from 6900XT to 290X because I need to return the 6900XT.

**Comment by rlarrosa1** [comment]
Hello, 

Just to report that this bug seems to continue.

After the last update of opensuse 15.4, sddm has stopped starting automatically with the error in the journalctl "could not start Display server on vt 7".

All worked fine on opensuse 15.4 before the update.

Before giving that error, journalctl shows:

Adding new display on vt 7 ...
Loading theme configuration from ""
Display server starting...
Running: /usr/bin/X -nolisten tcp -auth /run/sddm/{dc1a0>

And three seconds latter it seems that X has stopped and sddm says :

sddm[1757]: Failed to read display number from pipe
sddm[1757]: Could not start Display server on vt 7

Doing:
systemctl restart display-manager.service

Worked fine, so it seems a race condition.

To solve it I have modified:
/usr/lib/X11/display-manager

To do:
startproc -p $PIDFILE $DISPLAYMANAGER $XDMOPTIONS || rc_failed
sleep 5
killall sddm
startproc -p $PIDFILE $DISPLAYMANAGER $XDMOPTIONS || rc_failed

And it works fine again.

It seems that X finishes after about 3 seconds after starting.
Between starting X and sddm several nvidia modules are loaded, it seems to me that those must be loaded before X, but it is only my opinion. Perhaps another solution is loading the modules before starting X.

Thanks,

Rafael

**Comment by nimalu** [comment]
Hi, 
same for me. Sometimes, sddm fails to start.

```
Mai 15 18:06:56 sddm[607]: Could not start Display server on vt 1
Mai 15 18:06:56 sddm[607]: Attempt 3 starting the Display server on vt 1 failed
Mai 15 18:06:56 sddm[607]: Display server stopping...
Mai 15 18:06:56 sddm[607]: Failed to read display number from pipe
Mai 15 18:06:56 sddm[607]: Running: /usr/bin/X -nolisten tcp -background none -seat seat0 vt1 -auth /var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae} -noreset -displayfd 17
Mai 15 18:06:56 sddm[607]: Adding cookie to "/var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae}"
Mai 15 18:06:56 sddm[607]: Display server starting...
Mai 15 18:06:54 sddm[607]: Attempt 2 starting the Display server on vt 1 failed
Mai 15 18:06:54 sddm[607]: Display server stopping...
Mai 15 18:06:54 sddm[607]: Failed to read display number from pipe
Mai 15 18:06:54 sddm[607]: Running: /usr/bin/X -nolisten tcp -background none -seat seat0 vt1 -auth /var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae} -noreset -displayfd 17
Mai 15 18:06:54 sddm[607]: Adding cookie to "/var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae}"
Mai 15 18:06:54 sddm[607]: Display server starting...
Mai 15 18:06:52 sddm[607]: Attempt 1 starting the Display server on vt 1 failed
Mai 15 18:06:52 sddm[607]: Display server stopping...
Mai 15 18:06:52 sddm[607]: Failed to read display number from pipe
Mai 15 18:06:52 sddm[607]: Running: /usr/bin/X -nolisten tcp -background none -seat seat0 vt1 -auth /var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae} -noreset -displayfd 17
Mai 15 18:06:52 sddm[607]: Adding cookie to "/var/run/sddm/{40c35fd6-3b66-4c4d-9a6f-f50d269929ae}"
Mai 15 18:06:52 sddm[607]: Display server starting...
Mai 15 18:06:52 sddm[607]: Loading theme configuration from ""
Mai 15 18:06:52 sddm[607]: Adding new display on vt 1 ...
Mai 15 18:06:52 sddm[607]: Logind interface found
Mai 15 18:06:52 sddm[607]: Starting...
Mai 15 18:06:52 sddm[607]: Initializing...
```

sddm version is 0.19.0-9

**Comment by user3287** [comment]
Same problem for me, just that systemctl restart display-manager.service doesn't get sddm to start and makes sddm omit the same error to journalctl -b.
My system is an old HP Server without an APU, but it also doesn't work on this system when I plug in a GPU (the GPU works in the BIOS) (Nvidida GPU, my AMD GPU doesn't fit)
(Gddm version is newest currently available in Ubuntu 22.04 repos)

**Comment by hovancik** [comment]
Same issue, restating sddm or display manager service does not help

```
Feb 09 12:05:35 sidpad systemd[1]: Started sddm.service - Simple Desktop Display Manager.
Feb 09 12:05:35 sidpad sddm[2388]: Initializing...
Feb 09 12:05:35 sidpad sddm[2388]: Starting...
Feb 09 12:05:35 sidpad sddm[2388]: Logind interface found
Feb 09 12:05:35 sidpad sddm[2388]: Adding new display...
Feb 09 12:05:35 sidpad sddm[2388]: Loaded empty theme configuration
Feb 09 12:05:35 sidpad sddm[2388]: Xauthority path: "/run/sddm/xauth_EkXnlL"
Feb 09 12:05:35 sidpad sddm[2388]: Using VT 1
Feb 09 12:05:35 sidpad sddm[2388]: Display server starting...
Feb 09 12:05:35 sidpad sddm[2388]: Writing cookie to "/run/sddm/xauth_EkXnlL"
Feb 09 12:05:35 sidpad sddm[2388]: Running: /usr/bin/X -nolisten tcp -background none -seat seat0 vt1 -auth /run/sddm/xauth_EkXnlL -noreset -displayfd 16
Feb 09 12:05:35 sidpad sddm[2388]: Failed to start display server process.
Feb 09 12:05:35 sidpad sddm[2388]: Attempt 1 starting the Display server on vt 1 failed
Feb 09 12:05:37 sidpad sddm[2388]: Tried to start Xorg before previous instance exited
Feb 09 12:05:37 sidpad sddm[2388]: Attempt 2 starting the Display server on vt 1 failed
Feb 09 12:05:39 sidpad sddm[2388]: Tried to start Xorg before previous instance exited
Feb 09 12:05:39 sidpad sddm[2388]: Attempt 3 starting the Display server on vt 1 failed
Feb 09 12:05:39 sidpad sddm[2388]: Could not start Display server on vt 1

```

  Installed: 0.20.0-2 Debian Sid

**Comment by corbolais** [comment]
Same symptom here.

When executing the Xorg startup:

`/usr/bin/X -nolisten tcp -auth /var/run/sddm/deadbeef -background none -noreset -displayfd 17 -seat seat0 vt1`

I'm getting: `malloc(): unaligned tcache chunk detected`

Running:
X.Org X Server 1.21.1.7
X Protocol Version 11, Revision 0
Current Operating System: Linux ubuntu 6.2.0-1018-lowlatency #18-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov 15 17:01:40 UTC 2023 x86_64
Kernel command line: BOOT_IMAGE=/vmlinuz-6.2.0-1018-lowlatency root=/dev/sda1 ro nomodeset vt.handoff=7
xorg-server 2:21.1.7-1ubuntu3.6 (For technical support please see http://www.ubuntu.com/support) 
Current version of pixman: 0.42.2

Edit: Seems rather related to this: https://ubuntuforums.org/showthread.php?t=2493459

**Comment by hovancik** [comment]
@corbolais I managed to fix this by reinstalling Xorg server.

**Comment by alanruttenberg** [comment]
Just bit me: Debian Bookworm, AMD system with 4090. Found another workaround:

```sh
sudo systemctl edit sddm.service
```
Add 

```
[Service]
ExecStartPre=/usr/bin/sleep 10
```
Ugly, but...
